From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Tue, 4 Aug 2026 19:56:06 +0200
Subject: mfd: ac100: Allow writes to the ADDA tuning registers

The writeable range ends at LINEOUT_CTRL (0x59), but the codec driver
writes ADDA_TUNE1 (0x5a) when powering the headphone amp, to enable
zero-crossover volume updates - that write has been silently failing
with -EIO ever since. ADDA_TUNE3 (0x5c) also holds the internal
oscillator enable, which jack detection is about to need.

Extend the range over all three tuning registers.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/mfd/ac100.c       | 2 +-
 include/linux/mfd/ac100.h | 2 ++
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/mfd/ac100.c b/drivers/mfd/ac100.c
index 111111111111..222222222222 100644
--- a/drivers/mfd/ac100.c
+++ b/drivers/mfd/ac100.c
@@ -32,7 +32,7 @@ static const struct regmap_range ac100_writeable_ranges[] = {
 	regmap_reg_range(AC100_ADC_DIG_CTRL, AC100_ADC_VOL_CTRL),
 	regmap_reg_range(AC100_HMIC_CTRL1, AC100_HMIC_STATUS),
 	regmap_reg_range(AC100_DAC_DIG_CTRL, AC100_DAC_MXR_GAIN),
-	regmap_reg_range(AC100_ADC_APC_CTRL, AC100_LINEOUT_CTRL),
+	regmap_reg_range(AC100_ADC_APC_CTRL, AC100_ADDA_TUNE3),
 	regmap_reg_range(AC100_ADC_DAP_L_CTRL, AC100_ADC_DAP_OPT),
 	regmap_reg_range(AC100_DAC_DAP_CTRL, AC100_DAC_DAP_OPT),
 	regmap_reg_range(AC100_ADC_DAP_ENA, AC100_DAC_DAP_ENA),
diff --git a/include/linux/mfd/ac100.h b/include/linux/mfd/ac100.h
index 111111111111..222222222222 100644
--- a/include/linux/mfd/ac100.h
+++ b/include/linux/mfd/ac100.h
@@ -81,6 +81,8 @@ struct ac100_dev {
 #define AC100_SPKOUT_CTRL		0x58
 #define AC100_LINEOUT_CTRL		0x59
 #define AC100_ADDA_TUNE1 		0x5a
+#define AC100_ADDA_TUNE2 		0x5b
+#define AC100_ADDA_TUNE3 		0x5c
 
 /* ADC digital audio processing (high pass filter & auto gain control */
 #define AC100_ADC_DAP_L_STA		0x80
-- 
Armbian

